### PR TITLE
fix: clean up tmp dir

### PR DESCRIPTION
**Pull request by yajo**
Until now, Copier always left a lot of garbage in the tmp dir.

When running in batch, this could become a big problem, as templates would fill up memory (`/tmp/` is usually mounted as tmpfs).

@moduon MT-3411

### CODE REVIEW

**Comment by sisp** [comment]
Pydantic can't validate the bound cleanup method. I've been wondering anyway why the `Worker` and `Subproject` classes should be Pydantic models as they don't parse external data, which is Pydantic's primary use case.

**Comment by yajo** [comment]
My pydantism is very rusty now... Do you have any clue on how to fix the test? 🤔

File: copier/main.py [hunk 1]
```diff
@@ -187,6 +188,7 @@ class Worker:
     unsafe: bool = False
 
     answers: AnswersMap = field(default_factory=AnswersMap, init=False)
+    _cleanup_hooks: List[Callable] = field(default_factory=list, init=False)
```
Comment (sisp) [inline copier/main.py:191]: I _think_ you can wrap the `Callable` type hint with Pydantic's `SkipValidation` container (https://docs.pydantic.dev/latest/usage/validators/#special-types):

```suggestion
    _cleanup_hooks: List[SkipValidation[Callable]] = field(default_factory=list, init=False)
```

Don't forget to import it from the `pydantic` package.

I haven't tried it though.

Comment (sisp) [reply copier/main.py:191]: Oh, wait a second. It's about serialization, not validation. 🤔

**Comment by sisp** [comment]
There is an occurrence of `{{ _copier_conf|to_json }}`, and it seems `_copier_conf` now contains this cleanup callable. I believe

https://github.com/copier-org/copier/blob/a9e7b6ba3c9ae3fca6606eb8ae22e7da52b2d69d/copier/main.py#L285

adds `_cleanup_hooks` to the dict which can't be serialized to JSON. I guess it would work if you explicitly deleted it from the dict.

File: copier/main.py (hunk 1 above)
Comment (sisp) [reply copier/main.py:191]: I think this was wrong, so I'm resolving this conversation as invalid.

**Comment by sisp** [comment]
I took the liberty to fix the problem here, but for some reason GitHub doesn't show the commit 34d429fffda4e539e37e8c1478742a5fb51a01fc in https://github.com/copier-org/copier/tree/cleanup in this PR. :thinking:

**Comment by sisp** [comment]
Okay, not sure what went wrong, but I amended the commit without change to get a new commit hash and force-pushed again, now GitHub picked it up.

**Review by yajo — COMMENTED** [review]
Great, thanks!

**Review by sisp — APPROVED** [review]
LGTM! :tada:

The only minor hiccup I have is accessing the `Subproject._cleanup` method in the cached property `Worker.subproject`, but I think it's fine for now and would probably require an architectural change to avoid it.

I've pulled the latest changes in `master`, especially the changes from #1156, and adapted the `current_worker.run_copy()` call to use the context manager as well.

**Pull request merged**